kv/txn_coord_sender.go

```diff
@@ -600,7 +600,25 @@ func (tc *TxnCoordSender) heartbeat(id string, trace *tracer.Trace, ctx context.
 	// so grab a copy before unlocking.
 	txn := txnMeta.txn
 	tc.Unlock()
+
+	ba := roachpb.BatchRequest{}
+	ba.Timestamp = tc.clock.Now()
+	ba.CmdID = ba.GetOrCreateCmdID(ba.Timestamp.WallTime)
+	ba.Txn = txn.Clone()
+
 	if !proceed {
+		// Actively abort the transaction and its intents since we assume it's abandoned
+		ab := &roachpb.EndTransactionRequest{
+			Commit:  false,
+			Intents: txnMeta.intents(),
```

Comment:
you'll need to lock around here and atomically remove from `tc.txns`. Something like

```go
var intents []roachpb.Intent
tc.Lock()
if txnMeta, ok := tc.txns[id]; ok {
  delete(tc.txns, id)
  intents := txnMeta.intents()
}
tc.Unlock()